### ISSUE TITLE
Memory leak in v2.0.21

### ISSUE THREAD

**Issue by alanhamlett** [unbit/uwsgi#2537]
Upgraded uWSGI from `2.0.17.1` to `2.0.21` and started getting alarms in production after about an hour of traffic. Nothing else changed, and the leak goes away if I downgrade uWSGI back to `2.0.17.1`. I guess that indicates a memory leak was introduced somewhere in the changes between those two versions.

I also noticed worker segfaults with both versions, maybe `2.0.17.1` is cleaning up but `2.0.21` isn't?

```
...The work of process 1418021 is done. Seeya!
!!! uWSGI process 1418021 got Segmentation Fault !!!
*** backtrace of 1418021 ***
/usr/local/bin/uwsgi(uwsgi_backtrace+0x2e) [0x55a5e3b6de5e]
/usr/local/bin/uwsgi(uwsgi_segfault+0x27) [0x55a5e3b6e247]
/lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7f77b7d66090]
/opt/app/current/venv/lib/python3.8/site-packages/greenlet/_greenlet.cpython-38-x86_64-linux-gnu.so(_ZN24ThreadState_DestroyNoGIL19DestroyQueueWithGILEPv+0x230) [0x7f77af58d640]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x1c97f3) [0x7f77b80de7f3]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyEval_EvalFrameDefault+0x9e3) [0x7f77b7f8ab53]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x8006b) [0x7f77b7f9506b]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(_PyObject_CallMethodId+0x116) [0x7f77b81be9e6]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(+0x192038) [0x7f77b80a7038]
/lib/x86_64-linux-gnu/libpython3.8.so.1.0(Py_FinalizeEx+0x35) [0x7f77b80a85b5]
/usr/local/bin/uwsgi(uwsgi_plugins_atexit+0x71) [0x55a5e3b6c101]
/lib/x86_64-linux-gnu/libc.so.6(+0x468a7) [0x7f77b7d698a7]
/lib/x86_64-linux-gnu/libc.so.6(on_exit+0) [0x7f77b7d69a60]
/usr/local/bin/uwsgi(+0x425f5) [0x55a5e3b205f5]
/usr/local/bin/uwsgi(simple_goodbye_cruel_world+0x5f) [0x55a5e3b6d84f]
/usr/local/bin/uwsgi(+0x8f88c) [0x55a5e3b6d88c]
/usr/local/bin/uwsgi(uwsgi_close_request+0x5cf) [0x55a5e3b213cf]
/usr/local/bin/uwsgi(simple_loop_run+0xd0) [0x55a5e3b69b40]
/usr/local/bin/uwsgi(simple_loop+0x14) [0x55a5e3b69914]
/usr/local/bin/uwsgi(uwsgi_ignition+0x2a4) [0x55a5e3b6e5a4]
/usr/local/bin/uwsgi(uwsgi_worker_run+0x276) [0x55a5e3b72d06]
/usr/local/bin/uwsgi(uwsgi_run+0x454) [0x55a5e3b73284]
/usr/local/bin/uwsgi(+0x3ee84) [0x55a5e3b1ce84]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f77b7d47083]
/usr/local/bin/uwsgi(_start+0x2e) [0x55a5e3b1cebe]
*** end of backtrace ***
```

Edit: the leak is still present in `2.0.17.1` it just takes longer to trigger alarms so I didn't notice it before upgrading uWSGI. What initially triggered the memory leak is upgrading some Python libraries, which started that segfault above.

**Comment by dcere** [comment]
In my case it started when moving two projects from Ubuntu `20` to `22` (Python `3.8.10` to Python `3.10.6`).
One of them is having these segmentation faults every time a worker is restarted (even though it has not done any work).

Downgrading greenlet from `2.0.2` to `1.1.3` seems to fix that case.

**Comment by killershotpy** [comment]
what about the standard means within the python environment for clearing memory?
for example: https://docs.python.org/3/library/gc.html

**Comment by siddhantGupta04** [comment]
@dcere How did you figure out it was the **greenlet** that was causing the memory leak?

in my case, only a certain pod is getting restarted due to memory leak, while other pods remain healthy.

**Comment by ssill2** [comment]
I'm seeing a similar issue in 2.0.24.   I'm currently running a test with 2.0.25.1 to see if the problem still exists.  We also upgraded to python 3.12 at the same time.

**Comment by ssill2** [comment]
> @dcere How did you figure out it was the **greenlet** that was causing the memory leak?
> 
> in my case, only a certain pod is getting restarted due to memory leak, while other pods remain healthy.

same for me.  we have about 10 different pods that get deployed, it's only the one with uwsgi in it.  my java pods are fine lol   I'm trying to debug this for my python team.

**Comment by ssill2** [comment]
in my case, I believe my python brethren switched to uwsgi from gunicorn because gunicorn didn't support chunked-encoding.   I'm that's the extent of my familiarity with using uwsgi, but it seems to be the path common even the most simple requests, e.g. the pod healthcheck endpoint which does nothing more than return `true`

**Comment by ssill2** [comment]
they're using django, is the thread manager something implemented at the django level.  Sorry, I'm a java EE guy lol so I don't generally have these problems, but I'm trying to lend a hand in helping resolve this.

**Comment by ssill2** [comment]
or do you mean
https://pypi.org/project/threadmanager/

**Comment by ssill2** [comment]
I'm not 100% convinced that the problem I'm seeing is a uwsgi thing either.

**Comment by ssill2** [comment]
I just ran a test with the latest uwsgi version, 2.0.25.1  and I still see the issue.  the mem usage in the pod definitely climbed when I added a specific endpoint in the application and this gives me a place to look at to see if it's something in that endpoint that causes the problem.  The call to the endpoint takes about 4.9-5.2 seconds to service because of the data being fetched, so they are on the heavy side.

**Comment by killershotpy** [comment]
> I just ran a test with the latest uwsgi version, 2.0.25.1 and I still see the issue. the mem usage in the pod definitely climbed when I added a specific endpoint in the application and this gives me a place to look at to see if it's something in that endpoint that causes the problem. The call to the endpoint takes about 4.9-5.2 seconds to service because of the data being fetched, so they are on the heavy side.

No, I'm talking about a reasonable thread manager solution, self-written, not some library, libraries are often insufficient and not flexible, and the thread utilizer solution takes no more than 30-40 lines of code.

If you want to diagnose threads and memory more accurately, you should use the objgraph library (https://pypi.org/project/objgraph/).

**Comment by D0wn3r** [comment]
some update about this ? I have a Django app running with an uwsgi and I need to reload workers because the ram increase for nothing. Even healthcheck calls. Maybe it has a link with this issue

**Comment by ssill2** [comment]
I'm solving this problem a different way.  I'm rewriting the entire app in java lol.   problem solved.

**Comment by fivard** [comment]
Still experience memory leaks with any uWSGI version after moving from python  3.8 to 3.12. Thanks to valgrind which showed it. 
@killershotpy How can I manage threads if my pods are eating memory without any requests to thread related code? Can you explain please?

**Comment by D0wn3r** [comment]
Gunicorn seems to manage RAM better. I did fast test yesterday, it doesn't keep RAM until a reboot of worker as uwsgi seems to do

**Comment by killershotpy** [comment]
> Still experience memory leaks with any uWSGI version after moving from python 3.8 to 3.12. Thanks to valgrind which showed it. [@killershotpy](https://github.com/killershotpy) How can I manage threads if my pods are eating memory without any requests to thread related code? Can you explain please?

please read the python documentation, the link I attached above.

Everything is explained in detail there.

I guess you don't understand the root difference between “server side” and “code side”.

In our case uwsgi is just a server that “holds” your compiled code (program), so all memory leaks should be looked for only inside your code, and uwsgi itself has nothing to do with memory leaks.

Earlier, I wrote that it incorrectly handles errors in another issue, but it does not relate to memory leaks.

In my applications, I managed to completely get rid of memory leaks only after I had exhaustively studied the basics of the python language and memory management in it.
There are a lot of “background” DBMS connections, as well as file descriptors that have not been cleaned up, for example, and these are the “memory leaks” in general.
